docs/design/secrets.md

```diff
@@ -0,0 +1,194 @@
+2.  [Specification for image/container standardization based on volumes](https://github.com/docker/docker/issues/9277)
+3.  [Kubernetes service account proposal](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297)
+4.  [Secret proposal for docker](https://github.com/docker/docker/pull/6075)
+5.  [Continuating of secret proposal for docker](https://github.com/docker/docker/pull/6697)
+
+## Analysis TODOs
+
+Collecting remaining TODOs here:
+
+1.  Determine how/whether secrets interact with service accounts; are namespaces enough for the
+    security scope?
+2.  Can/should my security context affect what secrets I have access to?
+3.  Should there be a way to express that a container which consumes a secret should be restarted
+    when the secret changes?
+
+## Proposed Design
+
+### Overview
+
+This design proposes a new `Secret` resource which is mounted into containers with a new volume
+type. The secrets volume will be backed by a volume plugin that does the actual work of fetching
+the secret and placing it on the filesystem to be mounted in the container. Secrets may consist
+of multiple files. For example, an SSH key pair. In order to remove the burden from the end user
+in specifying every file that a secret consists of, it should be possible to mount all files
+provided by a secret with a single ```VolumeMount``` entry in the container specification.
+
+### Secret API Resource
+
+A new resource for secrets will be added to the API:
+
+```go
+type Secret struct {
+    TypeMeta
+    ObjectMeta
+
+    // Keys in this map are the paths relative to the volume
+    // presented to a container for this secret data.
+    Data map[string][]byte
+}
+```
+
+A new REST API and registry interface will be added to accompany the `Secret` resource.  The
+default implementation of the registry will store `Secret` information in etcd.  Future registry
+implementations could store the `TypeMeta` and `ObjectMeta` fields in etcd and store the secret
+data in another data store entirely, or store the whole object in another data store.
+
+### Secret Volume Source
+
+A new `SecretSource` type of volume source will be added to the ```VolumeSource``` struct in the
+API:
+
+```go
+type VolumeSource struct {
+    // Other fields omitted
+
+    // SecretSource represents a secret that should be presented in a volume
+    SecretSource *SecretSource `json:"secret"`
+}
+
+type SecretSource struct {
+
+    Target ObjectReference
+}
+```
+
+### Secret Volume Plugin
+
+The secret volume plugin would implement the actual retrieval and laying down of secrets on the
+Kubelet's file system. Secrets may be stored in a special secret registry, as Docker volumes, LDAP
+etc. See [Issue #2030](https://github.com/GoogleCloudPlatform/kubernetes/issues/2030) for options.
+The implementation of this plugin is outside of the scope of this proposal. A default Etcd-based
+version will be provided out of the box, but different solutions may be appropriate based on
```

Comment:
@erictune I was crossing the streams here.  By etcd I was referring to the registry implementation in the API server.  There are two facets of the storage problem here:

1.  Storage backing the API server (example: you want the API for secrets backed by an HSM instead of an etcd directory)
2.  Storage of secrets on the node

I'll disentangle these two in my next pass.